# AI-Powered Customer Retention Intelligence Platform

# Notebook 2: Data Cleaning & Preprocessing

---

## Project Objective

The objective of this notebook is to clean, transform, and prepare the customer churn dataset for exploratory data analysis and machine learning.

Raw datasets often contain missing values, incorrect data types, duplicated records, inconsistent categories, and irrelevant features that can negatively impact model performance. This notebook ensures that the dataset is accurate, consistent, and suitable for predictive modeling.

---

## Business Context

Machine learning models are only as good as the quality of the data used to train them. Before predicting customer churn, it is essential to ensure that the customer information is complete, reliable, and properly formatted.

The preprocessing tasks performed in this notebook will establish a high-quality analytical dataset that supports accurate business insights and predictive analytics.

---

## Expected Outputs

By the end of this notebook we will have:

- Production-ready dataset
- Correct data types
- Missing values handled
- Duplicate records removed
- Standardized categorical values
- Encoded target variable
- Dataset prepared for exploratory analysis

In [ ]:
import pandas as pd
import numpy as np
import os
import warnings

warnings.filterwarnings("ignore")
pd.set_option("display.max_columns", None)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
PROJECT_PATH = "/content/drive/MyDrive/customer-churn-prediction"

RAW_DATA_PATH = os.path.join(PROJECT_PATH, "data", "raw")
PROCESSED_DATA_PATH = os.path.join(PROJECT_PATH, "data", "processed")
MODELS_PATH = os.path.join(PROJECT_PATH, "models")
FIGURES_PATH = os.path.join(PROJECT_PATH, "figures")
REPORTS_PATH = os.path.join(PROJECT_PATH, "reports")
EXPORTS_PATH = os.path.join(PROJECT_PATH, "exports")

In [ ]:
required_folders = [
    RAW_DATA_PATH,
    PROCESSED_DATA_PATH,
    MODELS_PATH,
    FIGURES_PATH,
    REPORTS_PATH,
    EXPORTS_PATH
]

for folder in required_folders:
    os.makedirs(folder, exist_ok=True)

print("Project folder structure verified.")

Project folder structure verified.


In [ ]:
# Load Dataset
df = pd.read_csv(
    os.path.join(RAW_DATA_PATH, "WA_Fn-UseC_-Telco-Customer-Churn.csv")
)

print(df.shape)

(7043, 21)


In [ ]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [ ]:
#Review Categories
for col in df.select_dtypes("object").columns:
    print(col)
    print(df[col].unique())
    print("=" * 50)

customerID
['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
gender
['Female' 'Male']
Partner
['Yes' 'No']
Dependents
['No' 'Yes']
PhoneService
['No' 'Yes']
MultipleLines
['No phone service' 'No' 'Yes']
InternetService
['DSL' 'Fiber optic' 'No']
OnlineSecurity
['No' 'Yes' 'No internet service']
OnlineBackup
['Yes' 'No' 'No internet service']
DeviceProtection
['No' 'Yes' 'No internet service']
TechSupport
['No' 'Yes' 'No internet service']
StreamingTV
['No' 'Yes' 'No internet service']
StreamingMovies
['No' 'Yes' 'No internet service']
Contract
['Month-to-month' 'One year' 'Two year']
PaperlessBilling
['Yes' 'No']
PaymentMethod
['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
TotalCharges
['29.85' '1889.5' '108.15' ... '346.45' '306.6' '6844.5']
Churn
['No' 'Yes']


## Duplicate Record Assessment

Duplicate observations may introduce bias into predictive models and distort statistical analysis.

This section identifies duplicate customer records and removes them if necessary.

In [ ]:
duplicate_count = df.duplicated().sum()

print(f"Duplicate Records Found: {duplicate_count}")

if duplicate_count > 0:
    df.drop_duplicates(inplace=True)
    print("Duplicate records removed.")
else:
    print("No duplicate records found.")

Duplicate Records Found: 0
No duplicate records found.


## Missing Value Assessment

Missing values reduce data quality and may negatively impact predictive performance. This section identifies incomplete observations before applying appropriate treatment strategies.

In [ ]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


## Data Type Validation

Incorrect data types may lead to inaccurate statistical calculations and machine learning errors.

This section validates and corrects the data types where necessary.

In [ ]:
df.dtypes

,0
customerID,object
gender,object
SeniorCitizen,int64
Partner,object
Dependents,object
tenure,int64
PhoneService,object
MultipleLines,object
InternetService,object
OnlineSecurity,object


In [ ]:
df["TotalCharges"] = pd.to_numeric(
    df["TotalCharges"],
    errors="coerce"
)

In [ ]:
df["TotalCharges"] = df["TotalCharges"].fillna(
    df["TotalCharges"].median()
)

In [ ]:
# Remove duplicates
df.drop_duplicates(inplace=True)

print(df.shape)

(7043, 21)


## Feature Selection

The customerID column is a unique identifier used only for customer tracking.

Since it contains no predictive information, it is removed to prevent unnecessary complexity during exploratory analysis and machine learning.

In [ ]:
if "customerID" in df.columns:
    df.drop(columns=["customerID"], inplace=True)
    print("customerID removed successfully.")
else:
    print("customerID already removed.")

customerID removed successfully.


In [ ]:
print(df.columns.tolist())

['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']


In [ ]:
# Categorical Variable Validation
for col in df.select_dtypes("object").columns:
    print(col)
    print(df[col].unique())

gender
['Female' 'Male']
Partner
['Yes' 'No']
Dependents
['No' 'Yes']
PhoneService
['No' 'Yes']
MultipleLines
['No phone service' 'No' 'Yes']
InternetService
['DSL' 'Fiber optic' 'No']
OnlineSecurity
['No' 'Yes' 'No internet service']
OnlineBackup
['Yes' 'No' 'No internet service']
DeviceProtection
['No' 'Yes' 'No internet service']
TechSupport
['No' 'Yes' 'No internet service']
StreamingTV
['No' 'Yes' 'No internet service']
StreamingMovies
['No' 'Yes' 'No internet service']
Contract
['Month-to-month' 'One year' 'Two year']
PaperlessBilling
['Yes' 'No']
PaymentMethod
['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn
['No' 'Yes']


In [ ]:
#Data Cleaning Report

print("=" * 60)
print("DATA CLEANING REPORT")
print("=" * 60)

print(f"Rows                 : {df.shape[0]}")
print(f"Columns              : {df.shape[1]}")
print(f"Missing Values       : {df.isnull().sum().sum()}")
print(f"Duplicate Records    : {df.duplicated().sum()}")

print("\nData Types")
print(df.dtypes.value_counts())

print("\nChurn Distribution")
print(df["Churn"].value_counts())

print("\nDataset Status")
print("Ready for Exploratory Data Analysis")

DATA CLEANING REPORT
Rows                 : 7043
Columns              : 20
Missing Values       : 0
Duplicate Records    : 22

Data Types
object     16
int64       2
float64     2
Name: count, dtype: int64

Churn Distribution
Churn
No     5174
Yes    1869
Name: count, dtype: int64

Dataset Status
Ready for Exploratory Data Analysis


# Final Dataset Validation

After completing all preprocessing tasks, we perform a final statistical review of the cleaned dataset.

This summary verifies the structure, distribution, and completeness of both numerical and categorical variables before the dataset is used for exploratory data analysis and machine learning.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   MultipleLines     7043 non-null   object 
 7   InternetService   7043 non-null   object 
 8   OnlineSecurity    7043 non-null   object 
 9   OnlineBackup      7043 non-null   object 
 10  DeviceProtection  7043 non-null   object 
 11  TechSupport       7043 non-null   object 
 12  StreamingTV       7043 non-null   object 
 13  StreamingMovies   7043 non-null   object 
 14  Contract          7043 non-null   object 
 15  PaperlessBilling  7043 non-null   object 
 16  PaymentMethod     7043 non-null   object 


In [ ]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
gender,7043,2,Male,3555,NaN,NaN,NaN,NaN,NaN,NaN,NaN
SeniorCitizen,7043.0,NaN,NaN,NaN,0.162147,0.368612,0.0,0.0,0.0,0.0,1.0
Partner,7043,2,No,3641,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Dependents,7043,2,No,4933,NaN,NaN,NaN,NaN,NaN,NaN,NaN
tenure,7043.0,NaN,NaN,NaN,32.371149,24.559481,0.0,9.0,29.0,55.0,72.0
PhoneService,7043,2,Yes,6361,NaN,NaN,NaN,NaN,NaN,NaN,NaN
MultipleLines,7043,3,No,3390,NaN,NaN,NaN,NaN,NaN,NaN,NaN
InternetService,7043,3,Fiber optic,3096,NaN,NaN,NaN,NaN,NaN,NaN,NaN
OnlineSecurity,7043,3,No,3498,NaN,NaN,NaN,NaN,NaN,NaN,NaN
OnlineBackup,7043,3,No,3088,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
# Data Cleaning Report
print("="*60)
print("DATA CLEANING REPORT")
print("="*60)

print(f"Rows                 : {df.shape[0]}")
print(f"Columns              : {df.shape[1]}")
print(f"Missing Values       : {df.isnull().sum().sum()}")
print(f"Duplicate Records    : {df.duplicated().sum()}")

print("\nData Types")
print(df.dtypes.value_counts())

print("\nDataset Status")
print("READY FOR EXPLORATORY DATA ANALYSIS")

DATA CLEANING REPORT
Rows                 : 7043
Columns              : 20
Missing Values       : 0
Duplicate Records    : 22

Data Types
object     16
int64       2
float64     2
Name: count, dtype: int64

Dataset Status
READY FOR EXPLORATORY DATA ANALYSIS


In [ ]:
# Remove Duplicate Records | DO NOT TOUCH

duplicate_count = df.duplicated().sum()

print(f"Duplicate Records Found: {duplicate_count}")

df.drop_duplicates(inplace=True)

print(f"Dataset Shape After Removal: {df.shape}")

print(f"Remaining Duplicates: {df.duplicated().sum()}")

Duplicate Records Found: 22
Dataset Shape After Removal: (7021, 20)
Remaining Duplicates: 0


In [ ]:
#Export Clean Dataset
df.to_csv(
    os.path.join(PROCESSED_DATA_PATH, "customer_churn_clean.csv"),
    index=False
)

print(" customer_churn_clean.csv exported successfully.")

 customer_churn_clean.csv exported successfully.


# Notebook Summary

This notebook transformed the raw customer churn dataset into a production-ready analytical dataset suitable for machine learning.

The preprocessing workflow included duplicate removal, missing value treatment, data type correction, categorical standardization, target variable encoding, and dataset validation.

The cleaned dataset produced in this notebook will serve as the primary data source for the next phase of the project, **Exploratory Data Analysis (EDA)**, where customer behavior, churn patterns, and business insights will be investigated.

## Key Findings

• The raw dataset initially contained **7,043 customer records**.

• A total of **22 duplicate records** were identified and removed, resulting in **7,021 unique customer records**.

• Missing values in the **TotalCharges** feature were identified and successfully imputed using the median.

• The **TotalCharges** column was converted to a numeric data type.

• The **customerID** identifier column was removed because it does not contribute to predictive modeling.

• The cleaned dataset was exported as **customer_churn_clean.csv** and is ready for exploratory data analysis.